File: agent-gui/src-tauri/src/plugin_runtime/cache.rs
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

use super::types::CachedJsonEnvelope;

pub fn read_cached_json<T: serde::de::DeserializeOwned>(path: &Path) -> Option<T> {
    let raw = fs::read_to_string(path).ok()?;
    let envelope: CachedJsonEnvelope<T> = serde_json::from_str(&raw).ok()?;
    let now_ms = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .ok()?
        .as_millis() as u64;
    let ttl_ms = envelope.ttl_hours.saturating_mul(3_600_000);
    if now_ms.saturating_sub(envelope.fetched_at_ms) > ttl_ms {
        return None;
    }
    Some(envelope.payload)
}

pub fn read_stale_cached_json<T: serde::de::DeserializeOwned>(path: &Path) -> Option<T> {
    let raw = fs::read_to_string(path).ok()?;
    let envelope: CachedJsonEnvelope<T> = serde_json::from_str(&raw).ok()?;
    Some(envelope.payload)
}

pub fn write_cached_json<T: serde::Serialize>(
    path: &Path,
    payload: &T,
    ttl_hours: u64,
) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let fetched_at_ms = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|e| e.to_string())?
        .as_millis() as u64;
    let envelope = CachedJsonEnvelope {
        fetched_at_ms,
        ttl_hours,
        payload,
    };
    let raw = serde_json::to_string_pretty(&envelope).map_err(|e| e.to_string())?;
    fs::write(path, raw).map_err(|e| e.to_string())
}

File: agent-gui/electron/plugin-runtime/cache.mjs
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

export function readCachedJson(path) {
  try {
    const raw = readFileSync(path, "utf8");
    const envelope = JSON.parse(raw);
    const nowMs = Date.now();
    const ttlMs = Number(envelope.ttlHours ?? 0) * 3_600_000;
    if (nowMs - Number(envelope.fetchedAtMs ?? 0) > ttlMs) {
      return null;
    }
    return envelope.payload;
  } catch {
    return null;
  }
}

export function readStaleCachedJson(path) {
  try {
    const raw = readFileSync(path, "utf8");
    return JSON.parse(raw).payload;
  } catch {
    return null;
  }
}

export function writeCachedJson(path, payload, ttlHours) {
  mkdirSync(dirname(path), { recursive: true });
  const envelope = {
    fetchedAtMs: Date.now(),
    ttlHours,
    payload,
  };
  writeFileSync(path, `${JSON.stringify(envelope, null, 2)}\n`, "utf8");
}
